plugins/newsdownloader.koplugin/http_utilities.lua

```diff
@@ -0,0 +1,126 @@
+local logger = require("logger")
+local http = require("socket.http")
+local socketutil = require("socketutil")
+local socket_url = require("socket.url")
+local socket = require("socket")
+local ltn12 = require("ltn12")
+
+local NewsHelpers = {
+
+}
+
+local max_redirects = 5; --prevent infinite redirects
+
+-- Get URL content
+function NewsHelpers:getUrlContent(url, timeout, maxtime, redirectCount)
+    logger.dbg("getUrlContent(", url, ",", timeout, ",", maxtime, ",", redirectCount, ")")
+    if not redirectCount then
+        redirectCount = 0
+    elseif redirectCount == max_redirects then
+        error("EpubDownloadBackend: reached max redirects: ", redirectCount)
+    end
+
+    if not timeout then timeout = 10 end
+    logger.dbg("timeout:", timeout)
+
+    local sink = {}
+    local parsed = socket_url.parse(url)
+    socketutil:set_timeout(timeout, maxtime or 30)
+    local request = {
+        url     = url,
+        method  = "GET",
+        sink    = maxtime and socketutil.table_sink(sink) or ltn12.sink.table(sink),
+    }
+    logger.dbg("request:", request)
+    local code, headers, status = socket.skip(1, http.request(request))
+    socketutil:reset_timeout()
+    logger.dbg("After http.request")
+    local content = table.concat(sink) -- empty or content accumulated till now
+    logger.dbg("type(code):", type(code))
+    logger.dbg("code:", code)
+    logger.dbg("headers:", headers)
+    logger.dbg("status:", status)
+    logger.dbg("#content:", #content)
+
+    if code == socketutil.TIMEOUT_CODE or
+       code == socketutil.SSL_HANDSHAKE_CODE or
+       code == socketutil.SINK_TIMEOUT_CODE
+    then
+        logger.warn("request interrupted:", code)
+        return false, code
+    end
+    if headers == nil then
+        logger.warn("No HTTP headers:", code, status)
+        return false, "Network or remote server unavailable"
+    end
+    if not code or string.sub(code, 1, 1) ~= "2" then -- all 200..299 HTTP codes are OK
+        if code and code > 299 and code < 400  and headers and headers.location then -- handle 301, 302...
+           local redirected_url = headers.location
+           local parsed_redirect_location = socket_url.parse(redirected_url)
+           if not parsed_redirect_location.host then
+             parsed_redirect_location.host = parsed.host
+             parsed_redirect_location.scheme = parsed.scheme
```

Comment:
(4, 3, 2 spaces ! Indentation gets lower as we go deeper ? that's quite a new style :)

Reply:
Lol, imagine if that were like actually the style and not just a mistake? XD